presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroupManager.java

```diff
@@ -0,0 +1,265 @@
+    {
+        ResourceGroupId group;
+        try {
+            group = selectGroup(queryExecution.getSession());
+        }
+        catch (PrestoException e) {
+            queryExecution.fail(e);
+            return;
+        }
+        createGroupIfNecessary(group, queryExecution.getSession(), executor);
+        groups.get(group).run(queryExecution);
+    }
+
+    @Override
+    public void addConfigurationManagerFactory(ResourceGroupConfigurationManagerFactory factory)
+    {
+        if (configurationManagerFactories.putIfAbsent(factory.getName(), factory) != null) {
+            throw new IllegalArgumentException(format("Resource group configuration manager '%s' is already registered", factory.getName()));
+        }
+    }
+
+    @Override
+    public void loadConfigurationManager()
+            throws Exception
+    {
+        if (RESOURCE_GROUPS_CONFIGURATION.exists()) {
+            Map<String, String> properties = new HashMap<>(loadProperties(RESOURCE_GROUPS_CONFIGURATION));
+
+            String configurationManagerName = properties.remove(CONFIGURATION_MANAGER_PROPERTY_NAME);
+            checkArgument(!isNullOrEmpty(configurationManagerName),
+                    "Resource groups configuration %s does not contain %s", RESOURCE_GROUPS_CONFIGURATION.getAbsoluteFile(), CONFIGURATION_MANAGER_PROPERTY_NAME);
+
+            setConfigurationManager(configurationManagerName, properties);
+        }
+        else {
+            setConfigurationManager(BUILTIN_RESOURCE_GROUP_MANAGER, ImmutableMap.of());
+        }
+    }
+
+    @VisibleForTesting
+    public void setConfigurationManager(String name, Map<String, String> properties)
+    {
+        requireNonNull(name, "name is null");
+        requireNonNull(properties, "properties is null");
+
+        log.info("-- Loading resource group configuration manager --");
+
+        ResourceGroupConfigurationManagerFactory configurationManagerFactory = configurationManagerFactories.get(name);
+        checkState(configurationManagerFactory != null, "Resource group configuration manager %s is not registered", name);
+
+        ResourceGroupConfigurationManager configurationManager = configurationManagerFactory.create(ImmutableMap.copyOf(properties), () -> memoryPoolManager);
+        this.configurationManager.set(configurationManager);
+
+        log.info("-- Loaded resource group configuration manager %s --", name);
+    }
+
+    private static Map<String, String> loadProperties(File file)
+            throws Exception
+    {
+        requireNonNull(file, "file is null");
+
+        Properties properties = new Properties();
+        try (FileInputStream in = new FileInputStream(file)) {
+            properties.load(in);
+        }
+        return fromProperties(properties);
+    }
+
+    @PreDestroy
+    public void destroy()
+    {
+        refreshExecutor.shutdownNow();
+    }
+
+    @PostConstruct
+    public void start()
+    {
+        if (started.compareAndSet(false, true)) {
```

Comment:
Yes, it's just not legal to submit queries until it's set (I added a check for that)